### PR TITLE
Fix array arguments on Symfony process

### LINKED ISSUE #87: [TypeError] Symfony\Component\Process\Process::__construct(): Argument #1 ($command) must be of type array, string given
Happens during install with `composer` in version `2.3.0` (tested on PHP 8.1):

```
  ....
  - Installing menatwork/contao-multicolumnwizard-bundle (3.5.2): Extracting archive
  - Installing terminal42/contao-changelanguage (3.2.9): Extracting archive
  - Installing terminal42/contao-easy_themes (2.2.11): Extracting archive
  - Installing terminal42/dc_multilingual (4.2.2): Extracting archive
  - Installing toflar/psr6-symfony-http-cache-store (3.0.1): Extracting archive
    Install of codefog/contao-haste failed
    Install of terminal42/contao-easy_themes failed
    Install of terminal42/dc_multilingual failed
In Process.php line 143:
                                                                               
  [TypeError]                                                                  
  Symfony\Component\Process\Process::__construct(): Argument #1 ($command) mu  
  st be of type array, string given, called in /vendor/con  
  tao-community-alliance/composer-plugin/src/UserFilesLocator.php on line 99   
                                                                               
Exception trace:
  at phar:///composer.phar/vendor/symfony/process/Process.php:143
 Symfony\Component\Process\Process->__construct() at /vendor/contao-community-alliance/composer-plugin/src/UserFilesLocator.php:99
 ContaoCommunityAlliance\Composer\Plugin\UserFilesLocator->getPathFromConsole() at /vendor/contao-community-alliance/composer-plugin/src/UserFilesLocator.php:60
 ContaoCommunityAlliance\Composer\Plugin\UserFilesLocator->locate() at /vendor/contao-community-alliance/composer-plugin/src/Installer/AbstractModuleInstaller.php:234
 ContaoCommunityAlliance\Composer\Plugin\Installer\AbstractModuleInstaller->getFilesRoot() at /vendor/contao-community-alliance/composer-plugin/src/Installer/AbstractModuleInstaller.php:117
 ContaoCommunityAlliance\Composer\Plugin\Installer\AbstractModuleInstaller->ContaoCommunityAlliance\Composer\Plugin\Installer\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:28
 React\Promise\FulfilledPromise->then() at phar:///composer.phar/vendor/react/promise/src/Promise.php:134
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:42
 React\Promise\FulfilledPromise->done() at phar:///composer.phar/vendor/react/promise/src/Promise.php:135
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:42
 React\Promise\FulfilledPromise->done() at phar:///composer.phar/vendor/react/promise/src/Promise.php:66
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:42
 React\Promise\FulfilledPromise->done() at phar:///composer.phar/vendor/react/promise/src/Promise.php:135
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:42
 React\Promise\FulfilledPromise->done() at phar:///composer.phar/vendor/react/promise/src/Promise.php:135
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/FulfilledPromise.php:42
 React\Promise\FulfilledPromise->done() at phar:///composer.phar/vendor/react/promise/src/Promise.php:135
 React\Promise\Promise::React\Promise\{closure}() at phar:///composer.phar/vendor/react/promise/src/Promise.php:168
 React\Promise\Promise->settle() at phar:///composer.phar/vendor/react/promise/src/Promise.php:231
 React\Promise\Promise::React\Promise\{closure}() at n/a:n/a
 call_user_func() at phar:///composer.phar/src/Composer/Util/ProcessExecutor.php:321
 Composer\Util\ProcessExecutor->countActiveJobs() at phar:///composer.phar/src/Composer/Util/Loop.php:98
 Composer\Util\Loop->wait() at phar:///composer.phar/src/Composer/Installer/InstallationManager.php:496
 Composer\Installer\InstallationManager->waitOnPromises() at phar:///composer.phar/src/Composer/Installer/InstallationManager.php:469
 Composer\Installer\InstallationManager->executeBatch() at phar:///composer.phar/src/Composer/Installer/InstallationManager.php:390
 Composer\Installer\InstallationManager->downloadAndExecuteBatch() at phar:///composer.phar/src/Composer/Installer/InstallationManager.php:282
 Composer\Installer\InstallationManager->execute() at phar:///composer.phar/src/Composer/Installer.php:759
 Composer\Installer->doInstall() at phar:///composer.phar/src/Composer/Installer.php:281
 Composer\Installer->run() at phar:///composer.phar/src/Composer/Command/InstallCommand.php:137
 Composer\Command\InstallCommand->execute() at phar:///composer.phar/vendor/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at phar:///composer.phar/vendor/symfony/console/Application.php:1015
 Symfony\Component\Console\Application->doRunCommand() at phar:///composer.phar/vendor/symfony/console/Application.php:299
 Symfony\Component\Console\Application->doRun() at phar:///composer.phar/src/Composer/Console/Application.php:334
 Composer\Console\Application->doRun() at phar:///composer.phar/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at phar:///composer.phar/src/Composer/Console/Application.php:130
 Composer\Console\Application->run() at phar:///composer.phar/bin/composer:83
 require() at /composer.phar:29
install [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-suggest] [--no-dev] [--no-autoloader] [--no-progress] [--no-install] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--] [<packages>...]
```

### PR DESCRIPTION
**Pull request by bytehead**
Fixes #87

### CODE REVIEW

**Review by Toflar — APPROVED** [review]
This was changed in 4.2 so we don't need a BC layer here.

**Review by Toflar — CHANGES_REQUESTED** [review]
But the arguments need to be separated, no?

**Comment by bytehead** [comment]
Could possibly, I'll check.

**Comment by bytehead** [comment]
Changed in 9d5d93f accordingly.

**Review by Toflar — APPROVED** [review]

**Comment by fritzmg** [comment]
This will break compatibility with previous composer versions (and thus possibly also the Contao Manager - not sure though @aschempp).

**Comment by bytehead** [comment]
@fritzmg You're right, Composer itself ships before 2.3.0 with `symfony/process` in `2.8.x`. 
I've added a fallback in deed078.

File: src/UserFilesLocator.php [hunk 1]
```diff
@@ -96,7 +97,14 @@ private function getConsolePath()
      */
     private function getPathFromConsole()
     {
-        $console = new Process($this->getConsolePath() . ' debug:container --parameter=contao.upload_path');
+        $arguments = [$this->getConsolePath(), 'debug:container', '--parameter=contao.upload_path'];
+
+        // Fallback to string argument if Composer version is lower than 2.3.0 (see #87)
+        if (version_compare(Composer::getVersion(), '2.3.0', '<')) {
+            $arguments = $this->getConsolePath() . ' debug:container --parameter=contao.upload_path';
```
Comment (Toflar) [inline src/UserFilesLocator.php]:
```suggestion
            $arguments = implode(' ', $arguments);
```

Comment (bytehead) [reply src/UserFilesLocator.php]: Nice 💪

Comment (bytehead) [reply src/UserFilesLocator.php]: Adjusted in 169f383.

**Review by Toflar — APPROVED** [review]

**Review by fritzmg — APPROVED** [review]
Btw. this is how the Contao Manager Bundle's Composer plugin checks for it:

https://github.com/contao/contao/blob/fd1a75351dfb8a88dd92982b6a83ffd1a54518f3/manager-bundle/src/Composer/ScriptHandler.php#L80-L83

But checking the composer version is fine too :)

**Comment by aschempp** [comment]
I think we should go with the `setCommandline` check. The `symfony/process` version in the Contao Manager is different from what the pre-build `composer.phar` contains, since we only support PHP 7.1+ by default. Checking for `setCommandline` should handle all use cases.

**Comment by bytehead** [comment]
I don't get the correlation between the argument type hinting and `setCommandline`? 🤔

**Comment by fritzmg** [comment]
> I don't get the correlation between the argument type hinting and `setCommandline`? 🤔

The `symfony/process` version that introduced support for the array argument also introduced that method.

**Comment by leofeyer** [comment]
We should merge this ASAP, because Contao can no longer be installed with Composer 2.3.

**Review by leofeyer — APPROVED** [review]

**Pull request merged**